warehouse/query-core/src/main/java/datawave/query/jexl/visitors/RewriteNullFunctionsVisitor.java

```diff
@@ -0,0 +1,485 @@
+                    children.add(child);
+                }
+            }
+            JexlNodes.children(node, children.toArray(new JexlNode[0]));
+            rebuiltMultiFieldedFunction = false;
+        }
+        
+        return data;
+    }
+    
+    /**
+     * Rewrite <code>filter:isNull</code> and <code>filter:isNotNull</code> functions to the more efficient <code>FIELD == null</code> or
+     * <code>!(FIELD == null)</code>
+     *
+     * @param node
+     *            a function node
+     * @param data
+     *            an object
+     * @return the rewritten node, or the original node if no rewrite was required
+     */
+    @Override
+    public Object visit(ASTFunctionNode node, Object data) {
+        
+        FunctionJexlNodeVisitor visitor = new FunctionJexlNodeVisitor();
+        node.jjtAccept(visitor, null);
+        
+        if (visitor.namespace().equals(EVAL_PHASE_FUNCTION_NAMESPACE)) {
+            if (visitor.name().equals(IS_NULL)) {
+                JexlNode rewritten = rewriteFilterFunction(visitor, false);
+                JexlNodes.replaceChild(node.jjtGetParent(), node, rewritten);
+            } else if (visitor.name().equals(IS_NOT_NULL)) {
+                JexlNode rewritten = rewriteFilterFunction(visitor, true);
+                JexlNodes.replaceChild(node.jjtGetParent(), node, rewritten);
+            }
+        }
+        
+        return data;
+    }
+    
+    /**
+     * Given the output of a {@link FunctionJexlNodeVisitor}, produce a rewritten filter function
+     *
+     * @param visitor
+     *            a {@link FunctionJexlNodeVisitor}
+     * @param negated
+     *            a flag indicating this method should negate the built node
+     * @return the rewritten function
+     */
+    private JexlNode rewriteFilterFunction(FunctionJexlNodeVisitor visitor, boolean negated) {
+        JexlNode args = visitor.args().get(0);
+        if (args instanceof ASTIdentifier) {
+            // single fielded case
+            if (negated) {
+                return buildIsNotNullNode((ASTIdentifier) args);
+            } else {
+                return buildIsNullNode((ASTIdentifier) args);
+            }
+        } else if (args instanceof ASTOrNode) {
+            // multi fielded case
+            rebuiltMultiFieldedFunction = true;
+            List<JexlNode> children = new ArrayList<>(args.jjtGetNumChildren());
+            for (JexlNode child : JexlNodes.children(args)) {
+                child = JexlASTHelper.dereference(child);
+                if (child instanceof ASTIdentifier) {
+                    if (negated) {
+                        children.add(buildIsNotNullNode((ASTIdentifier) child));
+                    } else {
+                        children.add(buildIsNullNode((ASTIdentifier) child));
+                    }
+                } else {
+                    throw new IllegalStateException("Encountered an unexpected state while rewriting IsNotNull functions");
```

Comment:
I'd like to see this exception message be a little more descriptive, for example, what kind of node is child if it's not an ASTIdentifier?

Reply:
Fixed in next commit